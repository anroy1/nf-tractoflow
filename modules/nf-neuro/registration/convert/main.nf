process REGISTRATION_CONVERT {
    tag "$meta.id"
    label 'process_single'

    container "freesurfer/freesurfer:7.4.1"
    containerOptions "--env FSLOUTPUTTYPE='NIFTI_GZ'"

    input:
    tuple val(meta), path(affine), path(deform), path(source), path(target), path(fs_license)

    output:
    tuple val(meta), path("*.{txt,lta,mat,dat}"), emit: affine_transform
    tuple val(meta), path("*.{nii,nii.gz,mgz,m3z}"), emit: deform_transform
    path "versions.yml"           , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"

    //For arguments definition, lta_convert -h
    def invert = task.ext.invert ? "--invert" : ""
    def source_geometry_affine = "$source" ? "--src " + "$source" : ""
    def target_geometry_affine = "$target" ? "--trg " + "$target" : ""
    def in_format_affine = task.ext.in_format_affine ? "--in" + task.ext.in_format_affine + " " + "$affine" : "--inlta " + "$affine"
    def out_format_affine = task.ext.out_format_affine ? "--out" + task.ext.out_format_affine : "--outitk"

    //For arguments definition, mri_warp_convert -h
    def source_geometry_deform = "$source" ? "--insrcgeom " + "$source" : ""
    def in_format_deform = task.ext.in_format_deform ? "--in" + task.ext.in_format_deform + " " + "$deform" : "--inras " + "$deform"
    def out_format_deform = task.ext.out_format_deform ? "--out" + task.ext.out_format_deform : "--outitk"
    def downsample = task.ext.downsample ? "--downsample" : ""

    """
    export ITK_GLOBAL_DEFAULT_NUMBER_OF_THREADS=$task.cpus
    export OMP_NUM_THREADS=1
    export OPENBLAS_NUM_THREADS=1

    cp $fs_license \$FREESURFER_HOME/license.txt

    declare -A affine_dictionnary=( ["--outlta"]="lta" \
                                    ["--outfsl"]="mat" \
                                    ["--outmni"]="xfm" \
                                    ["--outreg"]="dat" \
                                    ["--outniftyreg"]="txt" \
                                    ["--outitk"]="txt" \
                                    ["--outvox"]="txt" )

    ext_affine=\${affine_dictionnary[${out_format_affine}]}

    declare -A deform_dictionnary=( ["--outm3z"]="m3z" \
                                    ["--outfsl"]="nii.gz" \
                                    ["--outlps"]="nii.gz" \
                                    ["--outitk"]="nii.gz" \
                                    ["--outras"]="nii.gz" \
                                    ["--outvox"]="mgz" )

    ext_deform=\${deform_dictionnary[${out_format_deform}]}

    lta_convert ${invert} ${source_geometry_affine} ${target_geometry_affine} ${in_format_affine} ${out_format_affine} ${prefix}__affine_warp.\${ext_affine}
    mri_warp_convert ${source_geometry_deform} ${downsample} ${in_format_deform} ${out_format_deform}  ${prefix}__deform_warp.\${ext_deform}

    rm \$FREESURFER_HOME/license.txt

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        Freesurfer: 7.4.1
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"

    """
    lta_convert -h
    mri_warp_convert -h

    touch ${prefix}__affine_transform.txt
    touch ${prefix}__deform_transform.nii.gz

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        Freesurfer: 7.4.1
    END_VERSIONS
    """
}
