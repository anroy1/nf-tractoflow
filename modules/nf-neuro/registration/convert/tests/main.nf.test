nextflow_process {

    name "Test Process REGISTRATION_CONVERT"
    script "../main.nf"
    process "REGISTRATION_CONVERT"

    tag "modules"
    tag "modules_nfcore"
    tag "registration"
    tag "registration/convert"

    tag "subworkflows"
    tag "subworkflows/load_test_data"


    setup {
        run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
            script "../../../../../subworkflows/nf-neuro/load_test_data/main.nf"
            process {
                """
                input[0] = Channel.from( [ "freesurfer.zip" , "freesurfer_reslice.zip" , "freesurfer_transforms.zip"] )
                input[1] = "test.load-test-data"
                """
            }
        }
    }

    test("registration - convert - default") {

        config "./nextflow_default.config"

        when {
            process {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        freesurfer: it.simpleName == "freesurfer"
                        reslice: it.simpleName == "freesurfer_reslice"
                        transforms: it.simpleName == "freesurfer_transforms"
                    }
                ch_transforms = ch_split_test_data.transforms.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/fs_affine.lta"),
                        file("\${test_data_directory}/fs_deform.nii.gz"),
                    ]
                }
                ch_reslice = ch_split_test_data.reslice.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/t1_reslice.nii.gz"),
                        []
                    ]
                }
                ch_freesurfer = ch_split_test_data.freesurfer.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/license.txt")
                    ]
                }
                input[0] = ch_transforms
                    .join(ch_reslice)
                    .join(ch_freesurfer)
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("registration - convert - fsants") {

        config "./nextflow_fsants.config"

        when {
            process {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        freesurfer: it.simpleName == "freesurfer"
                        reslice: it.simpleName == "freesurfer_reslice"
                        transforms: it.simpleName == "freesurfer_transforms"
                    }
                ch_transforms = ch_split_test_data.transforms.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/fs_affine.lta"),
                        file("\${test_data_directory}/fs_deform.nii.gz")
                    ]
                }
                ch_reslice = ch_split_test_data.reslice.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/t1_reslice.nii.gz"),
                        file("\${test_data_directory}/fa_reslice.nii.gz")
                    ]
                }
                ch_freesurfer = ch_split_test_data.freesurfer.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/license.txt")
                    ]
                }
                input[0] = ch_transforms
                    .join(ch_reslice)
                    .join(ch_freesurfer)
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("registration - convert - fslfs") {

        config "./nextflow_fslfs.config"

        when {
            process {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        freesurfer: it.simpleName == "freesurfer"
                        reslice: it.simpleName == "freesurfer_reslice"
                        transforms: it.simpleName == "freesurfer_transforms"
                    }
                ch_transforms = ch_split_test_data.transforms.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/fsl_affine.mat"),
                        file("\${test_data_directory}/fsl_deform.nii.gz")
                    ]
                }
                ch_reslice = ch_split_test_data.reslice.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/t1_reslice.nii.gz"),
                        file("\${test_data_directory}/fa_reslice.nii.gz")
                    ]
                }
                ch_freesurfer = ch_split_test_data.freesurfer.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/license.txt")
                    ]
                }
                input[0] = ch_transforms
                    .join(ch_reslice)
                    .join(ch_freesurfer)
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    file(process.out.affine_transform.get(0).get(1)).name,
                    process.out.deform_transform,
                    process.out.versions
                ).match() }
            )
        }
    }

    test("registration - convert - fslants") {

        config "./nextflow_fslants.config"

        when {
            process {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        freesurfer: it.simpleName == "freesurfer"
                        reslice: it.simpleName == "freesurfer_reslice"
                        transforms: it.simpleName == "freesurfer_transforms"
                    }
                ch_transforms = ch_split_test_data.transforms.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/fsl_affine.mat"),
                        file("\${test_data_directory}/fsl_deform.nii.gz")
                    ]
                }
                ch_reslice = ch_split_test_data.reslice.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/t1_reslice.nii.gz"),
                        file("\${test_data_directory}/fa_reslice.nii.gz")
                    ]
                }
                ch_freesurfer = ch_split_test_data.freesurfer.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/license.txt")
                    ]
                }
                input[0] = ch_transforms
                    .join(ch_reslice)
                    .join(ch_freesurfer)
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("registration - convert - antsfs") {

        config "./nextflow_antsfs.config"

        when {
            process {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        freesurfer: it.simpleName == "freesurfer"
                        reslice: it.simpleName == "freesurfer_reslice"
                        transforms: it.simpleName == "freesurfer_transforms"
                    }
                ch_transforms = ch_split_test_data.transforms.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/ants_affine.txt"),
                        file("\${test_data_directory}/ants_deform.nii.gz")
                    ]
                }
                ch_reslice = ch_split_test_data.reslice.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/t1_reslice.nii.gz"),
                        file("\${test_data_directory}/fa_reslice.nii.gz")
                    ]
                }
                ch_freesurfer = ch_split_test_data.freesurfer.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/license.txt")
                    ]
                }
                input[0] = ch_transforms
                    .join(ch_reslice)
                    .join(ch_freesurfer)
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    file(process.out.affine_transform.get(0).get(1)).name,
                    process.out.deform_transform,
                    process.out.versions
                ).match() }
            )
        }
    }
}
