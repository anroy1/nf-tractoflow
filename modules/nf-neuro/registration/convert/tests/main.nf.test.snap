{
    "registration - convert - fsants": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test__affine_warp.txt:md5,5f989a979be61faa578ad619377a8a07"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test__deform_warp.nii.gz:md5,69052e6226da946bad1f9466285cbb89"
                    ]
                ],
                "2": [
                    "versions.yml:md5,9912ec095965c1ff571f77b447c18f92"
                ],
                "affine_transform": [
                    [
                        {
                            "id": "test"
                        },
                        "test__affine_warp.txt:md5,5f989a979be61faa578ad619377a8a07"
                    ]
                ],
                "deform_transform": [
                    [
                        {
                            "id": "test"
                        },
                        "test__deform_warp.nii.gz:md5,69052e6226da946bad1f9466285cbb89"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,9912ec095965c1ff571f77b447c18f92"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-24T20:50:14.250271"
    },
    "registration - convert - default": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test__affine_warp.txt:md5,5f989a979be61faa578ad619377a8a07"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test__deform_warp.nii.gz:md5,69052e6226da946bad1f9466285cbb89"
                    ]
                ],
                "2": [
                    "versions.yml:md5,9912ec095965c1ff571f77b447c18f92"
                ],
                "affine_transform": [
                    [
                        {
                            "id": "test"
                        },
                        "test__affine_warp.txt:md5,5f989a979be61faa578ad619377a8a07"
                    ]
                ],
                "deform_transform": [
                    [
                        {
                            "id": "test"
                        },
                        "test__deform_warp.nii.gz:md5,69052e6226da946bad1f9466285cbb89"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,9912ec095965c1ff571f77b447c18f92"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-24T20:50:09.870665"
    },
    "registration - convert - antsfs": {
        "content": [
            "test__affine_warp.lta",
            [
                [
                    {
                        "id": "test"
                    },
                    "test__deform_warp.nii.gz:md5,599fbe3a6b85d61c0c67cea8be4972b7"
                ]
            ],
            [
                "versions.yml:md5,9912ec095965c1ff571f77b447c18f92"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-24T20:50:44.315777"
    },
    "registration - convert - fslants": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test__affine_warp.txt:md5,074e8ac5777a91ba0808cd58c5a0cc44"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test__deform_warp.nii.gz:md5,94feb8f0e648256eaa5ae0a47e5702c6"
                    ]
                ],
                "2": [
                    "versions.yml:md5,9912ec095965c1ff571f77b447c18f92"
                ],
                "affine_transform": [
                    [
                        {
                            "id": "test"
                        },
                        "test__affine_warp.txt:md5,074e8ac5777a91ba0808cd58c5a0cc44"
                    ]
                ],
                "deform_transform": [
                    [
                        {
                            "id": "test"
                        },
                        "test__deform_warp.nii.gz:md5,94feb8f0e648256eaa5ae0a47e5702c6"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,9912ec095965c1ff571f77b447c18f92"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-24T20:50:39.752948"
    },
    "registration - convert - fslfs": {
        "content": [
            "test__affine_warp.lta",
            [
                [
                    {
                        "id": "test"
                    },
                    "test__deform_warp.nii.gz:md5,df432a27c586e57cad93af7509941cd4"
                ]
            ],
            [
                "versions.yml:md5,9912ec095965c1ff571f77b447c18f92"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-24T20:50:26.951153"
    }
}